### DISCUSSION TITLE
Is there any way to use the setSinkId() using multitrack?

### DISCUSSION THREAD

**Discussion by StefSchalks** [katspaugh/wavesurfer.js#2838] [Q&A] [answered]
### Wavesurfer.js version(s):
Beta

### Browser and operating system version(s):
Electron

### Description
I'm making a track mixer where you should be able to pre-listen your mix (before it goes live), but you might want to assign this on a different audio device. I was wondering if there is a possibility to use setSinkId() in a multitrack and if so, how can I do this?

Thanks in advanced.

**Comment by katspaugh** [comment]
You can set a sink id on wavesurfer's media element directly like this:

```
wavesurfer.getMediaElement().setSinkId(sinkId)
```

If you're asking about https://github.com/katspaugh/wavesurfer-multitrack, there you can set it on the [multitrack object](https://github.com/katspaugh/wavesurfer-multitrack/blob/main/src/multitrack.ts#L487-L491).

**Comment by StefSchalks** [comment]
Yeah I was talking about `wavesurfer-multitrack` but when I use 
```js
multitrack.setSinkId(sinkId);
```
I'm getting the following error:
```
Uncaught (in promise) DOMException: The operation could not be performed and was aborted
```
Do you know why this happens?

**Reply by StefSchalks** [reply]
@katspaugh sorry to ping you, but do you know why this could happen?

**Reply by katspaugh** [reply]
Can you paste your entire code where you're getting a device id and setting it?

**Reply by StefSchalks** [reply]
@katspaugh sure, here is my multitrack code
sure.
```
I removed the code because it's irrelevant
```

**Reply by katspaugh** [reply]
I meant the part where you set the sink id.

**Reply by StefSchalks** [reply]
Oh I removed that part sorry. Here's the right code (it's in the canplay function)
```js
// Page styles
  document.body.style.background = '#161313'
  document.body.style.color = '#fff'
  
  // Play/pause button
  const button = document.querySelector('#play')
  button.disabled = true
  multitrack.once('canplay', () => {
    // Get the wavesurfer player instance
    multitrack.setSinkId('default');
    multitrack.setTime(multitrack.tracks[0].endCue - 5);

    button.disabled = false
    button.onclick = () => {
      multitrack.isPlaying() ? multitrack.pause() : multitrack.play()
      button.textContent = multitrack.isPlaying() ? 'Pause' : 'Play'
    }
    
    // On spacebar press, play/pause
    document.body.onkeyup = (e) => {
      if (e.keyCode === 32) {
        multitrack.isPlaying() ? multitrack.pause() : multitrack.play()
        button.textContent = multitrack.isPlaying() ? 'Pause' : 'Play'
      }
    }
  })
```

**Reply by katspaugh** [reply]
I'm not sure if `"default"` is a legit sink id. The [MDN page](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/setSinkId) says you should find a device by iterating the device list like this:
```
const devices = await navigator.mediaDevices.enumerateDevices();
const audioDevice = devices.find((device) => device.kind === "audiooutput");
const audio = document.createElement("audio");
await audio.setSinkId(audioDevice.deviceId);
console.log(`Audio is being played on ${audio.sinkId}`);
```

**Reply by StefSchalks** [reply]
When I iterate through that list on another page, it gives me `'default'` as a valid option. (+ it works on another audio player just fine (not a wavesurfer one, just one created using `new Audio()`

![image](https://github.com/katspaugh/wavesurfer.js/assets/114923678/3ae5c669-849c-4a96-b9e9-e87262bfe19b)

**Reply by katspaugh** [reply]
I see. My best guess it might be related to the fact that wavesurfer’s audio element is not added to the DOM tree. I’ll try a few things and let you know if I get it working.

**Reply by katspaugh** [reply, marked as answer]
I've fixed it by setting an initial sinkId. See https://github.com/katspaugh/wavesurfer-multitrack/pull/3
